### DISCUSSION TITLE
Optimistic mutations and success side effects

### DISCUSSION THREAD

**Discussion by fzaninotto** [TanStack/query#2986] [Q&A]
Hi, and thanks for the awesome library.

I'm exploring optimistic mutations. I want to offer a specialized hook that wraps useMutation and provides optimistic updates. I followed the great doc at https://react-query.tanstack.com/guides/optimistic-updates, and it's working fine, but I have a problem with success side effects.

This is how I'd like my custom hook to be called:

```js
const mutation = useUpdate();
mutation.mutate({ resource, id, data }, { onSuccess: () => { /* redirections, UI changes, etc. */})
```

This is how the hook looks like:

```js
const useUpdate = () => useMutation(
   ({ resource, id, data }) => {/** perform the API call **/},
   {
     onMutate: /** snapshot & prefill the queryCache **/
     onError: /** rollback to snapshot **/
     onSettled: /** invalidate queries **/
  }
)
```

In a true optimistic UI, the call time `onSuccess` callback should be triggered right after the `onMutate` callback was fired, and not on success (if a redirection occurs after the mutation completes, it's too late). But I can't find a way to do it. 

Is there a way to read call-time mutationOptions inside the parameters of the options callbacks?

**Comment by TkDodo** [comment]
if you want to optimistically re-direct, why not do it right after the `mutate` call?

```
const mutation = useUpdate();
mutation.mutate({ resource, id, data })

redirect()
```
